Use modularity on EL8 (#1111)
theforeman/foreman-documentation: guides/common/modules/proc_configuring-repositories.adoc

include::proc_configuring-repositories-el.adoc[]

> ifdef::foreman-el[]

Reviewer: where is this `ifdef` closed? the one in line 35 *technically* belongs to the `ifdef` in line 19?

include::proc_configuring-repositories-el.adoc[]

+

> . If the Ruby 2.5 module has already been enabled, you need to perform a module reset.

Reviewer: How does the user know if it is already enabled? (`dnf module list --enabled`).

Author: So which order should we take?

A) Try to enable and describe steps if it fails:

* Enable foreman/katello module
* If that fails, check `dnf module list --enabled`
* Reset ruby and/or postgresql
* Try again

B) prevent failure and instead be more explicit:

* Check `dnf module list --enabled`
* Reset ruby and/or postgresql if enabled
* Enable foreman/katello module

Reviewer: Given this is the installation documentation (not upgrade), I'd go with "enable" with a "note: if this errors our because you already have blah enabled, you need to reset via `...`")

> . If the Ruby 2.5 module has already been enabled, a module reset will need to be performed.

Reviewer 2: ```suggestion
. If the Ruby 2.5 module has already been enabled, you need to perform a module reset.
```

Author: I copied what was already there for PostgreSQL 10, but updated now.

+
[options="nowrap" subs="+quotes,attributes"]
----
# {package-manager} module reset ruby
----

> . If the PostgreSQL 10 module has already been enabled, you need to perform a module reset.

Reviewer: Same as above, but additionally: if the user actually has PostgreSQL 10 installed and *data* in there, 12 will not be able to start with that data.

I am not sure if this file is also included in *upgrade* docs, but if it is, this would require additional clarification.

Author: See what I wrote in the commit message/PR description. Right now the upgrade docs completely rely on foreman-maintain. I have no clue how that's going to handle things. Or if that even works in upstream.

Reviewer: Right, I missed that.

I think f-m doesn't support upstream upgrades yet.

Author: Which also makes me wonder: reading https://docs.theforeman.org/nightly/Upgrading_and_Updating/index-katello.html it looks like it does install the new repositories but it doesn't talk about EL8 at all.

> . If the PostgreSQL 10 module has already been enabled, a module reset will need to be performed.

Reviewer 2: same as above.

----
# {package-manager} config-manager --set-enabled powertools
----

+
. Enable the Katello module:
+
[options="nowrap" subs="+quotes,attributes"]
----

> # {package-manager} module enable katello:el8

Reviewer: ```suggestion
# {package-manager} module enable katello:el8 pulpcore:el8
```

unless we decide to make katello depend on pulpcore, which we so far avoided.

Author: I was sure I checked if you could enable multiple modules at the same time on the CLI, but it looks like I didn't or I missed something.

Reviewer: You can.

----
endif::[]

=== In case module enablement fails

> If the module fails to enable, it can mean an incorrect module is enabled. List the enabled modules to find out which ones:

Author: I think my message didn't really come across that well so I'll give the background.

When you run `dnf module enable foreman:el8` (or katello, doesn't really matter) it can fail, but the message that dnf shows can be misleading. It shows warnings about postgresql, but the real reason can be that the Ruby 2.5 module is enabled. Thus users must run `dnf module list --enabled`. If there you see `ruby:2.5` (or really, any ruby module), you must follow the Ruby step. Same if you see the postgresql module. It can also be that both ruby and postgresql are enabled. In that case both steps must be followed.

That is what I'm trying to tell users in a short way. It really isn't that you're attempting to enable an incorrect module, but a module that was previously enabled that blocks enabling of the foreman/katello:el8 module.

> If the module fails to enable, it can mean an incorrect module is enabled. List the enabled modules to find out which ones:

Reviewer: One thing to consider (but I am not sure if that belongs in this paragraph or not), even if there are no problems, `dnf` produces a warning:

```console
# dnf module enable foreman:el8
Problems in request:
Modular dependency problems with Defaults:

 Problem: module foreman:el8:999920220314215004:6b5f9578.x86_64 requires module(postgresql:12), but none of the providers can be installed
  - module postgresql:10:8040020201214234117:9f9e2e7e.x86_64 conflicts with module(postgresql:12) provided by postgresql:12:8040020201220000948:9f9e2e7e.x86_64
  - module postgresql:12:8040020201220000948:9f9e2e7e.x86_64 conflicts with module(postgresql:10) provided by postgresql:10:8040020201214234117:9f9e2e7e.x86_64
  - module postgresql:10:8040020201214234117:9f9e2e7e.x86_64 conflicts with module(postgresql:12) provided by postgresql:12:8050020210609162852:b4937e53.x86_64
  - module postgresql:12:8050020210609162852:b4937e53.x86_64 conflicts with module(postgresql:10) provided by postgresql:10:8040020201214234117:9f9e2e7e.x86_64
  - module postgresql:10:8040020201214234117:9f9e2e7e.x86_64 conflicts with module(postgresql:12) provided by postgresql:12:8050020210624150605:b4937e53.x86_64
  - module postgresql:12:8050020210624150605:b4937e53.x86_64 conflicts with module(postgresql:10) provided by postgresql:10:8040020201214234117:9f9e2e7e.x86_64
  - module postgresql:10:8040020201214234117:9f9e2e7e.x86_64 conflicts with module(postgresql:12) provided by postgresql:12:8050020210716150548:b4937e53.x86_64
  - module postgresql:12:8050020210716150548:b4937e53.x86_64 conflicts with module(postgresql:10) provided by postgresql:10:8040020201214234117:9f9e2e7e.x86_64
  - module postgresql:10:8040020201214234117:9f9e2e7e.x86_64 conflicts with module(postgresql:12) provided by postgresql:12:8060020211110235049:d63f516d.x86_64
  - module postgresql:12:8060020211110235049:d63f516d.x86_64 conflicts with module(postgresql:10) provided by postgresql:10:8040020201214234117:9f9e2e7e.x86_64
  - module postgresql:10:8040020201214234117:9f9e2e7e.x86_64 conflicts with module(postgresql:12) provided by postgresql:12:8060020211213144215:d63f516d.x86_64
  - module postgresql:12:8060020211213144215:d63f516d.x86_64 conflicts with module(postgresql:10) provided by postgresql:10:8040020201214234117:9f9e2e7e.x86_64
  - conflicting requests
Dependencies resolved.
============================================================================
 Package          Architecture    Version            Repository        Size
============================================================================
Enabling module streams:
 foreman                          el8                                      
 postgresql                       12                                       
 ruby                             2.7                                      

Transaction Summary
============================================================================

Is this ok [y/N]: 
```

The resolution `dnf` comes up with is perfectly fine, but it contradicts the default modules and thus triggers the warning. How do we tell the users that *this* is OK, but others are not?

Especially as the output is almost identical:

```console
# dnf module enable foreman:el8
Last metadata expiration check: 0:03:49 ago on Wed Mar 16 07:34:09 2022.
Error: Problems in request:
Modular dependency problems:

 Problem: module foreman:el8:999920220314215004:6b5f9578.x86_64 requires module(postgresql:12), but none of the providers can be installed
  - module postgresql:10:8040020201214234117:9f9e2e7e.x86_64 conflicts with module(postgresql:12) provided by postgresql:12:8040020201220000948:9f9e2e7e.x86_64
  - module postgresql:12:8040020201220000948:9f9e2e7e.x86_64 conflicts with module(postgresql:10) provided by postgresql:10:8040020201214234117:9f9e2e7e.x86_64
  - module postgresql:10:8040020201214234117:9f9e2e7e.x86_64 conflicts with module(postgresql:12) provided by postgresql:12:8050020210609162852:b4937e53.x86_64
  - module postgresql:12:8050020210609162852:b4937e53.x86_64 conflicts with module(postgresql:10) provided by postgresql:10:8040020201214234117:9f9e2e7e.x86_64
  - module postgresql:10:8040020201214234117:9f9e2e7e.x86_64 conflicts with module(postgresql:12) provided by postgresql:12:8050020210624150605:b4937e53.x86_64
  - module postgresql:12:8050020210624150605:b4937e53.x86_64 conflicts with module(postgresql:10) provided by postgresql:10:8040020201214234117:9f9e2e7e.x86_64
  - module postgresql:10:8040020201214234117:9f9e2e7e.x86_64 conflicts with module(postgresql:12) provided by postgresql:12:8050020210716150548:b4937e53.x86_64
  - module postgresql:12:8050020210716150548:b4937e53.x86_64 conflicts with module(postgresql:10) provided by postgresql:10:8040020201214234117:9f9e2e7e.x86_64
  - module postgresql:10:8040020201214234117:9f9e2e7e.x86_64 conflicts with module(postgresql:12) provided by postgresql:12:8060020211110235049:d63f516d.x86_64
  - module postgresql:12:8060020211110235049:d63f516d.x86_64 conflicts with module(postgresql:10) provided by postgresql:10:8040020201214234117:9f9e2e7e.x86_64
  - module postgresql:10:8040020201214234117:9f9e2e7e.x86_64 conflicts with module(postgresql:12) provided by postgresql:12:8060020211213144215:d63f516d.x86_64
  - module postgresql:12:8060020211213144215:d63f516d.x86_64 conflicts with module(postgresql:10) provided by postgresql:10:8040020201214234117:9f9e2e7e.x86_64
  - conflicting requests
```

The only differences are in `Modular dependency problems with Defaults:` vs `Modular dependency problems:` in the initial message and the fact that there is no suggested resolution.

> If the module fails to enable, it can mean an incorrect module is enabled. List the enabled modules to find out which ones:

Reviewer 3: @ekohl I'm sorry, I misread that. In that case, I have a different solution. Hang on.

> If the module fails to enable, it can mean an incorrect module is enabled. List the enabled modules to find out which ones:

Reviewer 3: Intro paragraph after the heading: 
"If the XY*) module fails to enable, it can mean an incorrect module is enabled." + a sentence in the spirit "In that case, you have to resolve dependencies manually as follows." (or something like that to summarize what the user has to do about it next)

> If the module fails to enable, it can mean an incorrect module is enabled. List the enabled modules to find out which ones:

Author: Thanks @evgeni for sharing the explicit output. You can indeed see how similar they are and it's **very** easy for a user to misread the output (I know I had issues). It was my intent here to address that.

[options="nowrap" subs="+quotes,attributes"]
----
# dnf module list --enabled
----

==== Ruby

If the Ruby 2.5 module has already been enabled, you need to perform a module reset:
[options="nowrap" subs="+quotes,attributes"]
----
# dnf module reset ruby

> ----

Reviewer 3: This would be the second list item in the procedure with the heading dropped.

==== PostgreSQL

If the PostgreSQL 10 module has already been enabled, you need to perform a module reset:
[options="nowrap" subs="+quotes,attributes"]
----
# dnf module reset postgresql
----

When a database was previously created using PostgreSQL 10, an upgrade must be performed:

. Enable the module:
+
[options="nowrap" subs="+quotes,attributes"]
----
ifdef::katello[]
# dnf module enable katello:el8
endif::[]
ifndef::katello[]
# dnf module enable foreman:el8
endif::[]
----

+
. Install the PostgreSQL upgrade package:
+
[options="nowrap" subs="+quotes,attributes"]
----
# dnf install postgresql-upgrade
----

+
. Perform the upgrade:
+
[options="nowrap" subs="+quotes,attributes"]
----
# postgresql-setup --upgrade

> ----

Reviewer 3: This whole thing would be another description list item with a nested numbered list. Make sure it's actually nested ;)
